src/main/java/org/sagebionetworks/bridge/services/PermissionService.java

```diff
@@ -158,6 +172,120 @@ public void deletePermission(String appId, String guid) {
+                        Map<String, Set<AccessLevel>> entityTypeMap = additionalPermissions.getOrDefault(entityType, new HashMap<>());
+                        entityTypeMap.put(entityId, additionalAccessLevels);
+                        additionalPermissions.put(entityType, entityTypeMap);
+                    } else {
+                        Map<String, Set<AccessLevel>> entityTypeMap = additionalPermissions.getOrDefault(entityType, new HashMap<>());
+                        entityTypeMap.put(entityId, newPermissionEntityTypeMap.get(entityId));
+                        additionalPermissions.put(entityType, entityTypeMap);
+                    }
+                }
+            } else {
+                // the old permissions do not have any of the new so add them all
+                additionalPermissions.put(entityType, newPermissions.get(entityType));
+            }
+        }
+        
+        for (EntityType entityType : oldPermissions.keySet()) {
+            if (newPermissions.containsKey(entityType)) {
+                Map<String, Set<AccessLevel>> oldPermissionEntityTypeMap = newPermissions.get(entityType);
+                for (String entityId : oldPermissionEntityTypeMap.keySet()) {
+                    if (newPermissions.get(entityType).containsKey(entityId)) {
+                        Set<AccessLevel> newAccessLevels = newPermissions.get(entityType).get(entityId);
+                        Set<AccessLevel> removableAccessLevels = new HashSet<>();
+                        for (AccessLevel accessLevel : oldPermissionEntityTypeMap.get(entityId)) {
+                            if (!newAccessLevels.contains(accessLevel)) {
+                                removableAccessLevels.add(accessLevel);
+                            }
+                        }
+                        Map<String, Set<AccessLevel>> entityTypeMap = removablePermissions.getOrDefault(entityType, new HashMap<>());
+                        entityTypeMap.put(entityId, removableAccessLevels);
+                        removablePermissions.put(entityType, entityTypeMap);
+                    } else {
+                        Map<String, Set<AccessLevel>> entityTypeMap = removablePermissions.getOrDefault(entityType, new HashMap<>());
+                        entityTypeMap.put(entityId, oldPermissionEntityTypeMap.get(entityId));
+                        removablePermissions.put(entityType, entityTypeMap);
+                    }
+                }
+            } else {
+                removablePermissions.put(entityType, oldPermissions.get(entityType));
+            }
+        }
+        
+        List<Permission> currentPermissions = permissionDao.getPermissionsForUser(account.getAppId(), account.getId());
+        for (EntityType entityType : removablePermissions.keySet()) {
+            for (String entityId : removablePermissions.get(entityType).keySet()) {
+                for (AccessLevel accessLevel : removablePermissions.get(entityType).get(entityId)) {
```

Comment:
It might be wortwhile storing removablePermissions.get(entityType) in a temporary variable, so that you only have to call get() once. (Similarly in the other nested loops as well.)

Reply:
Updated.